packages/tailwindcss/src/ast.ts

```diff
@@ -256,11 +256,23 @@ export function optimizeAst(ast: AstNode[]) {
 
     // Rule
     else if (node.kind === 'rule') {
-      let copy = { ...node, nodes: [] }
-      for (let child of node.nodes) {
-        transform(child, copy.nodes, depth + 1)
+      // Rules with `&` as the selector should be flattened
+      if (node.selector === '&') {
+        for (let child of node.nodes) {
+          let nodes: AstNode[] = []
+          transform(child, nodes, depth + 1)
+          parent.push(...nodes)
+        }
+      }
```

Comment:
Just an optimization to flatten:
```css
.foo {
  & {
    color: red;
  }
}
```

Into:
```css
.foo {
  color: red;
}
```

Which is more relevant when introducing this new `@variant` you can use.